### PR TITLE
Added property ignoreHTTPStatus to be able to load from non-http source (i.e. file://)

### PR DESCRIPTION
**Pull request by leitzler**
Proposed fix for #7525 that still keeps the possibility to load from "file://".

### CODE REVIEW

**Comment by mrdoob** [comment]
How about checking for `this.readyState === 4` only?

```javascript
if ( this.readyState === 4 ) {

	if ( onLoad ) onLoad( response );

	scope.manager.itemEnd( url );

} else {

	if ( onError ) onError( event );

	scope.manager.itemError( url );

}
```

**Comment by leitzler** [comment]
Would be unwise. When using HTTP we must consider status & readyState to prevent issues like #7464 (and also to follow W3C spec http://www.w3.org/TR/XMLHttpRequest/).

Since non-HTTP protocols are not mentioned in the specification, browsers seems to implement this differently. I've seen different combinations of status & readyState in different browsers where all of them are OK (for example status = 0 & readyState = 3).

To support file:// status & readyState needs to be ignored, to support HTTP they both needs to be checked.

**Comment by mrdoob** [comment]
I see... How about renaming `ignoreHTTPstarus` to `strict` and setting it to `true` by default?

**Comment by leitzler** [comment]
Sounds fine to me, unless people think it has to do with javascript strict mode (http://www.w3schools.com/js/js_strict.asp). It's no reserved word so it should be ok to use.

I'll update the PR when I get home from work.

File: src/loaders/XHRLoader.js [hunk 1]
```diff
@@ -46,7 +48,7 @@ THREE.XHRLoader.prototype = {
 
 			THREE.Cache.add( url, response );
 
-			if ( this.status === 200 && this.readyState === 4 ) {
+			if ( ( this.status === 200 && this.readyState === 4 ) && scope.strict ) {
```
Comment (mrdoob) [inline src/loaders/XHRLoader.js]: I think the check should be...

```javascript
if ( scope.strict === false || ( this.status === 200 && this.readyState === 4 ) ) {
```

**Comment by mrdoob** [comment]
Actually... Maybe it would be simpler to do something like...

```javascript
var isLocal = url.substr( 0, 7 );
```

and...

```javascript
if ( isLocal === true || ( this.status === 200 && this.readyState === 4 ) ) {
```

**Comment by leitzler** [comment]
I wrote the check the way I did intentionally, with status & readyState grouped in parentheses (since it is one check) and having them first (since the normal thing to do when handling an XMLHttpRequest should be to follow the spec, and not to is an exception).

I guess you missed out `=== "file://"` in your isLocal example :)
That solution crossed my mind too, but I think there might be other cases that we won't cover then. Is it possible to use a data-URI? It would need some research anyway. And then again, the normal thing should be to follow specs, anything else is an exception, so I discarded the idea.

**Comment by leitzler** [comment]
Also, url could be "/path/to/file" so we would have to check the responseURL instead. And responseURL isn't available in all browser versions :-1:

**Comment by leitzler** [comment]
Just checked with Safari and loading using data URI, and it works!
```javascript
loader.load("data:application/json;charset=utf-8,%7B%7D",...
```
Safari then responds with `readyState === 4` and `status === 0` upon success.

**Comment by mrdoob** [comment]
> I wrote the check the way I did intentionally, with status & readyState grouped in parentheses (since it is one check) and having them first (since the normal thing to do when handling an XMLHttpRequest should be to follow the spec, and not to is an exception).

The check was wrong though. If `strict` was set to `false`, the condition would fail and we would be calling `onError()`.

> I guess you missed out === "file://" in your isLocal example :)

Yes 😇

> Also, url could be "/path/to/file" so we would have to check the responseURL instead. And responseURL isn't available in all browser versions :-1:

Meh...

Hmm, what about...

```javascript
if ( this.readyState === 4 ) {

	if ( this.status !== 200 ) {

		console.warn( 'THREE.XHRLoader: ...' );

	}

	if ( onLoad ) onLoad( response );

	scope.manager.itemEnd( url );

} else {

	if ( onError ) onError( event );

	scope.manager.itemError( url );

}
```

**Comment by mrdoob** [comment]
Hmm, that doesn't make sense actually.

Maybe something like this?

```javascript
switch( this.status ) {
	case 404:
		if ( onError ) onError( event );
		scope.manager.itemError( url );
		break;
	case 0:
		console.warn( 'THREE.XHRLoader: ', url, this. statusText );
		// note that I'm not adding a break here
	default:
		if ( onLoad ) onLoad( response );
		scope.manager.itemEnd( url );
}
```

**Comment by leitzler** [comment]
I understand that you prefer a solution that is transparent for both http and non http URI :smile: 

Since `status === 0` isn't even in the spec I guess we can treat it special. How about something more like:
```javascript
if ( this.status === 200 && this.readyState === 4) {
  // ok
} else if ( this.status === 0 ) {
    console.warn( '...');
  // handle as ok
} else {
  // error
}
```

This will follow specification for http requests, and still accept 0 as status (regardless of readyState, as I've seen it be other than 4 upon success for files).

Note that it is still important to check for `readyState` in http-requests, otherwise we can end up in several loads. One if we get an event:
- readyState = 2 `HEADERS_RECEIVED`, one or several if we then get:
- readyState = 3 `LOADING` and then when we finally get:
- readyState = 4 `DONE`.

**Comment by mrdoob** [comment]
> Since `status === 0` isn't even in the spec I guess we can treat it special. How about something more like:
```javascript
if ( this.status === 200 && this.readyState === 4) {
  // ok
} else if ( this.status === 0 ) {
    console.warn( '...');
  // handle as ok
} else {
  // error
}
```

Works for me! 😀

>Note that it is still important to check for `readyState` in http-requests, otherwise we can end up in several loads. One if we get an event:
- readyState = 2 `HEADERS_RECEIVED`, one or several if we then get:
- readyState = 3 `LOADING` and then when we finally get:
- readyState = 4 `DONE`.

I'm not sure about this... I think it makes sense to check the `readyState` value when listening the `onreadystatechange` event. But in this case we're listening the `load` event, so `readyState` should always be `4`?

**Comment by leitzler** [comment]
Good point about readyState. Updated again, sorry for the delay!

**Comment by mrdoob** [comment]
Perfect!

**Comment by mrdoob** [comment]
Many thanks!

**Pull request merged**